### DISCUSSION TITLE
High RAM Usage with Large Collection - Optimization Possibilities?

### DISCUSSION THREAD

**Discussion by ruben-quilez** [qdrant/qdrant#4826] [General]
### Environment

Qdrant version: v1.10.1
Deployment: Docker
Hardware specifications: CPU: Intel( i7-8700 CPU @ 3.20GHz. RAM: 126G

### Issue Description
I'm encountering high RAM usage with my Qdrant setup, consuming over 100GB of RAM across multiple collections. One collection, in particular, is significantly larger than the others. Here are the details of the largest collection:

```{
  "status": "green",
  "optimizer_status": {
    "error": "optimizations pending, awaiting update operation"
  },
  "indexed_vectors_count": 22038549,
  "points_count": 22043774,
  "segments_count": 58,
  "config": {
  "params": {
    "vectors": {
      "size": 1024,
      "distance": "Cosine"
    },
    "shard_number": 1,
    "replication_factor": 1,
    "write_consistency_factor": 1,
    "on_disk_payload": true
  },
  "hnsw_config": {
    "m": 16,
    "ef_construct": 100,
    "full_scan_threshold": 10000,
    "max_indexing_threads": 0,
    "on_disk": false
  },
  "optimizer_config": {
    "deleted_threshold": 0.2,
    "vacuum_min_vector_number": 1000,
    "default_segment_number": 0,
    "max_segment_size": null,
    "memmap_threshold": null,
    "indexing_threshold": 20000,
    "flush_interval_sec": 5,
    "max_optimization_threads": null
  },
  "wal_config": {
    "wal_capacity_mb": 32,
    "wal_segments_ahead": 0
  },
  "quantization_config": null
}}

```

### Questions

1. Is this level of RAM usage normal for a collection of this size? Are there any irregularities in the provided information that might explain the high memory consumption?
2. Are there any optimization strategies available to reduce RAM usage without significantly impacting performance?
3. If we implement optimizations to reduce RAM usage, what impact might this have on query response times?
4. Given the large number of indexed vectors (22,038,549) compared to the total points (22,043,774), is this discrepancy expected or could it indicate an issue?
5. The optimizer status shows "optimizations pending, awaiting update operation". Could this be contributing to the high RAM usage, and if so, how can we resolve this?
6. Are there any recommended best practices for managing collections of this size in terms of segmentation, HNSW configuration, or other parameters?
7. Could you provide guidance on how to interpret the memory usage in Qdrant? Is there a way to break down which components (e.g., HNSW index, WAL, payload, collection...) are consuming the most memory?

Any insights or recommendations would be greatly appreciated. Thank you for your time and assistance!

**Comment by ruben-quilez** [comment]
I almost forgot, this is my config.yaml:

```
log_level: DEBUG

storage:
  # Where to store all the data
  storage_path: ./storage

  # Where to store snapshots
  snapshots_path: ./snapshots

  # Where to store temporary files
  # If null, temporary snapshot are stored in: storage/snapshots_temp/
  temp_path: null

  # If true - point's payload will not be stored in memory.
  # It will be read from the disk every time it is requested.
  # This setting saves RAM by (slightly) increasing the response time.
  # Note: those payload values that are involved in filtering and are indexed - remain in RAM.
  on_disk_payload: true

  # Maximum number of concurrent updates to shard replicas
  # If `null` - maximum concurrency is used.
  update_concurrency: null

  # Write-ahead-log related configuration
  wal:
    # Size of a single WAL segment
    wal_capacity_mb: 32

    # Number of WAL segments to create ahead of actual data requirement
    wal_segments_ahead: 0

  # Normal node - receives all updates and answers all queries
  node_type: "Normal"

  # Listener node - receives all updates, but does not answer search/read queries
  # Useful for setting up a dedicated backup node
  # node_type: "Listener"

  performance:
    # Number of parallel threads used for search operations. If 0 - auto selection.
    max_search_threads: 0

    # Max number of threads (jobs) for running optimizations across all collections, each thread runs one job.
    # If 0 - have no limit and choose dynamically to saturate CPU.
    # Note: each optimization job will also use `max_indexing_threads` threads by itself for index building.
    max_optimization_threads: 0

    # CPU budget, how many CPUs (threads) to allocate for an optimization job.
    # If 0 - auto selection, keep 1 or more CPUs unallocated depending on CPU size
    # If negative - subtract this number of CPUs from the available CPUs.
    # If positive - use this exact number of CPUs.
    optimizer_cpu_budget: 0

    # Prevent DDoS of too many concurrent updates in distributed mode.
    # One external update usually triggers multiple internal updates, which breaks internal
    # timings. For example, the health check timing and consensus timing.
    # If null - auto selection.
    update_rate_limit: null

    # Limit for number of incoming automatic shard transfers per collection on this node, does not affect user-requested transfers.
    # The same value should be used on all nodes in a cluster.
    # Default is to allow 1 transfer.
    # If null - allow unlimited transfers.
    #incoming_shard_transfers_limit: 1

    # Limit for number of outgoing automatic shard transfers per collection on this node, does not affect user-requested transfers.
    # The same value should be used on all nodes in a cluster.
    # Default is to allow 1 transfer.
    # If null - allow unlimited transfers.
    #outgoing_shard_transfers_limit: 1

  optimizers:
    # The minimal fraction of deleted vectors in a segment, required to perform segment optimization
    deleted_threshold: 0.2

    # The minimal number of vectors in a segment, required to perform segment optimization
    vacuum_min_vector_number: 1000

    # Target amount of segments optimizer will try to keep.
    # Real amount of segments may vary depending on multiple parameters:
    #  - Amount of stored points
    #  - Current write RPS
    #
    # It is recommended to select default number of segments as a factor of the number of search threads,
    # so that each segment would be handled evenly by one of the threads.
    # If `default_segment_number = 0`, will be automatically selected by the number of available CPUs
    default_segment_number: 0

    # Do not create segments larger this size (in KiloBytes).
    # Large segments might require disproportionately long indexation times,
    # therefore it makes sense to limit the size of segments.
    #
    # If indexation speed have more priority for your - make this parameter lower.
    # If search speed is more important - make this parameter higher.
    # Note: 1Kb = 1 vector of size 256
    # If not set, will be automatically selected considering the number of available CPUs.
    max_segment_size_kb: null

    # Maximum size (in KiloBytes) of vectors to store in-memory per segment.
    # Segments larger than this threshold will be stored as read-only memmaped file.
    # To enable memmap storage, lower the threshold
    # Note: 1Kb = 1 vector of size 256
    # To explicitly disable mmap optimization, set to `0`.
    # If not set, will be disabled by default.
    memmap_threshold_kb: null

    # Maximum size (in KiloBytes) of vectors allowed for plain index.
    # Default value based on https://github.com/google-research/google-research/blob/master/scann/docs/algorithms.md
    # Note: 1Kb = 1 vector of size 256
    # To explicitly disable vector indexing, set to `0`.
    # If not set, the default value will be used.
    indexing_threshold_kb: 20000

    # Interval between forced flushes.
    flush_interval_sec: 5

    # Max number of threads (jobs) for running optimizations per shard.
    # Note: each optimization job will also use `max_indexing_threads` threads by itself for index building.
    # If null - have no limit and choose dynamically to saturate CPU.
    # If 0 - no optimization threads, optimizations will be disabled.
    max_optimization_threads: null

  # Default parameters of HNSW Index. Could be overridden for each collection or named vector individually
  hnsw_index:
    # Number of edges per node in the index graph. Larger the value - more accurate the search, more space required.
    m: 16

    # Number of neighbours to consider during the index building. Larger the value - more accurate the search, more time required to build index.
    ef_construct: 100

    # Minimal size (in KiloBytes) of vectors for additional payload-based indexing.
    # If payload chunk is smaller than `full_scan_threshold_kb` additional indexing won't be used -
    # in this case full-scan search should be preferred by query planner and additional indexing is not required.
    # Note: 1Kb = 1 vector of size 256
    full_scan_threshold_kb: 10000

    # Number of parallel threads used for background index building.
    # If 0 - automatically select.
    # Best to keep between 8 and 16 to prevent likelihood of building broken/inefficient HNSW graphs.
    # On small CPUs, less threads are used.
    max_indexing_threads: 0

    # Store HNSW index on disk. If set to false, index will be stored in RAM. Default: false
    on_disk: true

    # Custom M param for hnsw graph built for payload index. If not set, default M will be used.
    payload_m: null

  # Default shard transfer method to use if none is defined.
  # If null - don't have a shard transfer preference, choose automatically.
  # If stream_records, snapshot or wal_delta - prefer this specific method.
  # More info: https://qdrant.tech/documentation/guides/distributed_deployment/#shard-transfer-method
  shard_transfer_method: null

service:
  # Maximum size of POST data in a single request in megabytes
  max_request_size_mb: 32

  # Number of parallel workers used for serving the api. If 0 - equal to the number of available cores.
  # If missing - Same as storage.max_search_threads
  max_workers: 0

  # Host to bind the service on
  host: 0.0.0.0

  # HTTP(S) port to bind the service on
  http_port: 6333

  # gRPC port to bind the service on.
  # If `null` - gRPC is disabled. Default: null
  # Comment to disable gRPC:
  grpc_port: 6334

  # Enable CORS headers in REST API.
  # If enabled, browsers would be allowed to query REST endpoints regardless of query origin.
  # More info: https://developer.mozilla.org/en-US/docs/Web/HTTP/CORS
  # Default: true
  enable_cors: true

  # Enable HTTPS for the REST and gRPC API
  enable_tls: true

  # Check user HTTPS client certificate against CA file specified in tls config
  verify_https_client_certificate: false

  # Set an api-key.
  # If set, all requests must include a header with the api-key.
  # example header: `api-key: <API-KEY>`
  #
  # If you enable this you should also enable TLS.
  # (Either above or via an external service like nginx.)
  # Sending an api-key over an unencrypted channel is insecure.
  #
  # Uncomment to enable.
  # api_key: your_secret_api_key_here

  # Set an api-key for read-only operations.
  # If set, all requests must include a header with the api-key.
  # example header: `api-key: <API-KEY>`
  #
  # If you enable this you should also enable TLS.
  # (Either above or via an external service like nginx.)
  # Sending an api-key over an unencrypted channel is insecure.
  #
  # Uncomment to enable.
  # read_only_api_key: your_secret_read_only_api_key_here

  # Uncomment to enable JWT Role Based Access Control (RBAC).
  # If enabled, you can generate JWT tokens with fine-grained rules for access control.
  # Use generated token instead of API key.
  #
  # jwt_rbac: true

cluster:
  # Use `enabled: true` to run Qdrant in distributed deployment mode
  enabled: false

  # Configuration of the inter-cluster communication
  p2p:
    # Port for internal communication between peers
    port: 6335

    # Use TLS for communication between peers
    enable_tls: false

  # Configuration related to distributed consensus algorithm
  consensus:
    # How frequently peers should ping each other.
    # Setting this parameter to lower value will allow consensus
    # to detect disconnected nodes earlier, but too frequent
    # tick period may create significant network and CPU overhead.
    # We encourage you NOT to change this parameter unless you know what you are doing.
    tick_period_ms: 100


# Set to true to prevent service from sending usage statistics to the developers.
# Read more: https://qdrant.tech/documentation/guides/telemetry
telemetry_disabled: false


# TLS configuration.
# Required if either service.enable_tls or cluster.p2p.enable_tls is true.
tls:
  # Server certificate chain file
  cert: ./config/tls/cert.pem

  # Server private key file
  key: ./config/tls/key.pem

  # Certificate authority certificate file.
  # This certificate will be used to validate the certificates
  # presented by other nodes during inter-cluster communication.
  
  # If verify_https_client_certificate is true, it will verify
  # HTTPS client certificate
  
  # Required if cluster.p2p.enable_tls is true.
  ca_cert: ./config/tls/cacert.pem

  # TTL in seconds to reload certificate from disk, useful for certificate rotations.
  # Only works for HTTPS endpoints. Does not support gRPC (and intra-cluster communication).
  # If `null` - TTL is disabled.
  # cert_ttl: 3600

```

**Reply by rishabh1815769** [reply]
@ruben-quilez you can store your vectors and their index on the disk rather than storing them on RAM. check **on_disk** param for both.

**Reply by ruben-quilez** [reply]
Yes, I have changed both parameters in that collection to true, but I see that the RAM consumption has not decreased. Should an additional process have been executed?
I followed the documentation at https://qdrant.tech/documentation/concepts/collections/#update-collection-parameters
Current configuration of my collection:

```
{
  "params": {
    "vectors": {
      "size": 1024,
      "distance": "Cosine"
    },
    "shard_number": 1,
    "replication_factor": 1,
    "write_consistency_factor": 1,
    "on_disk_payload": true
  },
  "hnsw_config": {
    "m": 16,
    "ef_construct": 100,
    "full_scan_threshold": 10000,
    "max_indexing_threads": 0,
    "on_disk": true
  },
  "optimizer_config": {
    "deleted_threshold": 0.2,
    "vacuum_min_vector_number": 1000,
    "default_segment_number": 0,
    "max_segment_size": null,
    "memmap_threshold": null,
    "indexing_threshold": 20000,
    "flush_interval_sec": 5,
    "max_optimization_threads": null
  },
  "wal_config": {
    "wal_capacity_mb": 32,
    "wal_segments_ahead": 0
  },
  "quantization_config": null
}
```

**Reply by rishabh1815769** [reply]
Maybe the existing points are still in RAM and were not transferred to disk. Did you notice an increase in disk usage after updating the parameters?

**Reply by ruben-quilez** [reply]
Yes, they probably are still in RAM. I haven't noticed a significant increase in disk usage either. Why haven't the changes been applied despite the configuration indicating so?
Thanks!

**Reply by rishabh1815769** [reply]
Maybe you can wait for some time or try restarting your instance. If this doesn't work, re-creating the collection from scratch while providing **on_disk** params in advance would be the best option. We can raise a bug if this doesn't work. I come from a Qdrant version which did not have the functionality to update collection params so I am not sure why this is not working.

**Reply by ruben-quilez** [reply]
Yes, I've restarted the instance and even forced optimization of the collection, but it hasn't had any effect. I was hoping to avoid having to recreate the collection, as the process of inserting the vectors took days.
I've also seen that it's possible to recreate a collection from another one, but it indicates "WARNING:root:init_from is deprecated":

```
from qdrant_client import QdrantClient

client = QdrantClient(url="http://localhost:6333")

client.create_collection(
    collection_name="{collection_name}",
    vectors_config=models.VectorParams(size=100, distance=models.Distance.COSINE),
    init_from=models.InitFrom(collection="{from_collection_name}"),
)
```

**Reply by rishabh1815769** [reply]
Ohh ! Inserting the vectors should not take so long when you have around 22 million vectors. A couple of hours max. BTW how did you force the optimization of the collection?

**Reply by rishabh1815769** [reply]
Have you tried this method for parallel uploads and disabled indexing before inserting the points?

`qdrant.upload_points(
            "all_chains_invoices_embeddings_v3",
            points=batch,
            batch_size=8,
            parallel=16
        )`

**Reply by ruben-quilez** [reply]
For uploading the vectors, I'm using this library from LangChain:
https://api.python.langchain.com/en/latest/vectorstores/langchain_community.vectorstores.qdrant.Qdrant.html#langchain_community.vectorstores.qdrant.Qdrant.aadd_documents

```
@retry(tries=3, delay=2)
def load_vectordatabase(self, docs_chunks: tp.List[Document]) -> list:
    point_list = self._vector_store.add_documents(docs_chunks)
    return point_list
```

**Reply by ruben-quilez** [reply]
For forze optimice collection:

```
qdrant_client.update_collection(
    collection_name="collection_name",
    optimizer_config=models.OptimizersConfigDiff(),
)
```

**Comment by TerryT9** [comment]
set vector config **on_disk**=true will help to reduce the usage of memory.
[prefer-low-memory-footprint-with-high-speed-search](https://qdrant.tech/documentation/guides/optimize/#prefer-low-memory-footprint-with-high-speed-search)

**Comment by OmarEldeeb98** [comment]
what is the size of other collections ?
because I am facing the same issue of high memory usage.
Have you find a solution?
Thank you

**Reply by timvisee** [reply]
If you have a lot of collections, the memory optimization options (some suggested above) are significantly reduced.

@OmarEldeeb98 How many collections do you have?

**Reply by OmarEldeeb98** [reply]
@timvisee  currently I have 58 collections and memory usage 5.6GB, every collection has 2500 points with the following configurations 
`{
  "status": "green",
  "optimizer_status": "ok",
  "vectors_count": null,
  "indexed_vectors_count": 0,
  "points_count": 1557,
  "segments_count": 96,
  "config": {
    "params": {
      "vectors": {
        "size": 1536,
        "distance": "Cosine",
        "hnsw_config": null,
        "quantization_config": null,
        "on_disk": true,
        "datatype": null,
        "multivector_config": null
      },
      "shard_number": 12,
      "sharding_method": "auto",
      "replication_factor": 2,
      "write_consistency_factor": 1,
      "read_fan_out_factor": null,
      "on_disk_payload": true,
      "sparse_vectors": null
    },
    "hnsw_config": {
      "m": 16,
      "ef_construct": 100,
      "full_scan_threshold": 10000,
      "max_indexing_threads": 0,
      "on_disk": false,
      "payload_m": null
    },
    "optimizer_config": {
      "deleted_threshold": 0.2,
      "vacuum_min_vector_number": 1000,
      "default_segment_number": 0,
      "max_segment_size": null,
      "memmap_threshold": null,
      "indexing_threshold": 20000,
      "flush_interval_sec": 5,
      "max_optimization_threads": null
    },
    "wal_config": {
      "wal_capacity_mb": 32,
      "wal_segments_ahead": 0
    },
    "quantization_config": null,
    "strict_mode_config": {
      "enabled": false,
      "max_query_limit": null,
      "max_timeout": null,
      "unindexed_filtering_retrieve": null,
      "unindexed_filtering_update": null,
      "search_max_hnsw_ef": null,
      "search_allow_exact": null,
      "search_max_oversampling": null,
      "upsert_max_batchsize": null,
      "max_collection_vector_size_bytes": null,
      "read_rate_limit": null,
      "write_rate_limit": null,
      "max_collection_payload_size_bytes": null,
      "filter_max_conditions": null,
      "condition_max_size": null
    }
  },
  "payload_schema": {}
}`

**Reply by timvisee** [reply]
I highly recommend to use a single collection for everything instead, even if you have different users/tenants.

That will most definitely resolve the memory problems you're seeing.

Please see our documentation here, which describes this: https://qdrant.tech/documentation/guides/multiple-partitions/

**Reply by OmarEldeeb98** [reply]
@timvisee Thank you for fast reply
 if I switched `hnsw_config` to be on disk will it reduce memory usage significantly ?

**Reply by timvisee** [reply]
In this specific case, no, I do not expect it to have an effect. Because the number of points in your collection is relatively low, it currently does not have any HNSW at all.

I do now notice that you're having a high number of segments. 96 to be exact. Did you configure this explicitly, or was it chosen automatically?

**Reply by OmarEldeeb98** [reply]
@timvisee it was chosen automatically

**Reply by timvisee** [reply]
I expect it automatically picked 96 because you have a big machine.

Once again, I **really highly recommend** to switch to a single collection as described [here](https://qdrant.tech/documentation/guides/multiple-partitions/). That will be the only long-term/scalable solution.

But, if you want a temporary solution in the meanwhile, I can recommend to lower the number of segments. That should significantly reduce the memory usage for now.

Because you're having a low number of points, using 8 segments should be enough. If you're scaling to more than millions of point you'd have to adjust it though. My above suggested approach of a single collection does not suffer from this problem.

To change it, you'd have to call this on every collection:

```http
PATCH collections/my_collection
{
  "optimizers_config": {
    "default_segment_number": 8
  }
}
```

Each collection will then undergo background optimizations to switch down to 8 segments. Once all collections are green the background tasks are done.

**Reply by OmarEldeeb98** [reply]
@timvisee thanks 🙏🙏

**Reply by OmarEldeeb98** [reply]
@timvisee is there any reference mentions why many collections consume more resources than multitenancy?

**Reply by OmarEldeeb98** [reply]
@timvisee sorry for mentioning again, is there a list of data saved per collection that causes high memory for large number of qdrant?

**Reply by timvisee** [reply]
> As many as you want, but be aware that each collection requires additional resources.

We have this in our [FAQ](https://qdrant.tech/documentation/faq/qdrant-fundamentals/#collections), but it's quite sparse in terms of information.

The reason is the amount of components needed with one versus many collections. To give an example, the number of shards probably multiplies with the number of collections you have. Each shard has a WAL, which consumes memory to keep space for new incoming operations. There many more components like this. All together, this adds up. It makes having a single collection more efficient than having many.

**Comment by onestardao** [comment]
> Been following this thread with interest — and I think the deeper issue isn't *just* segment count or collection size.
> You’re bumping into what I call **“memory mirage”** in vector DBs:
>
> > *Where RAM spikes aren’t caused by the data itself, but the way it’s interpreted, buffered, and redundantly indexed during async ops, optimizer stages, and WAL interplay.*
>
> In this case, you're likely seeing high RAM due to the **"partial write, optimizer lag, and HNSW over-indexing triple punch"**:
>
> 1. **Pending optimizer cycles** keep full segment metadata and temporary trees in RAM even if `on_disk_payload=true` — meaning the RAM load isn’t cleanly flushable until after optim runs.
>
> 2. **Too many collections** = too many WALs, too many independent HNSW heaps, too many idle but memory-persistent config states.
>    (Timvise's advice on reducing collections is 🔥 but not granular enough for this edge.)
>
> 3. **Over-indexing**: If your `hnsw_ef_construct` and segment parallelism is too aggressive, you're burning RAM as if prepping for billion-scale search — even when you’re nowhere near that point size.

---

### Here's a diagnostic trick I use in my own framework:

```python
# Pseudo-code (actual implementation open-sourced under TXT OS):
def estimate_wasted_ram(qdrant_state):
    segments = qdrant_state["segments"]
    hnsw_nodes = segments * avg_ef_construct * 2
    payload_cache_size = num_points * avg_payload_size
    wal_reserve = collections * 2MB
    return hnsw_nodes * 80 + payload_cache_size + wal_reserve
```

I use this to visualize *semantic overhead* — not just raw storage. Because remember:

> **Qdrant doesn't store data. It stores possibilities.** And those chew RAM like no tomorrow if not tamed.

---

### 🛠 Recommended Fix:

* Set `default_segment_number: 1~4` for small-scale loads to limit fragmentation.
* Patch collections to use `optimizer_config.force_reindex=true` after bulk insert, so cleanup runs instantly.
* Consolidate small collections into a single tagged one (multi-tenancy via payload).
* If you’re async-writing: use staggered commits or inject wait-on-last semantics (`wait=true` on tail op).

---

> Bonus: I run a semantic layer on top of Qdrant that prevents this kind of RAM runaway via **predictive update throttling** — happy to share more if anyone’s interested.

Let me know if this helps — or if you’d like a script to profile this visually. I’ve been optimizing RAG memory behaviors all year.
